lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -740,8 +741,8 @@ defmodule Mix.Compilers.Elixir do
   end
 
   defp apply_warnings(sources, %{runtime_warnings: r_warnings, compile_warnings: c_warnings}) do
-    runtime_group = Enum.group_by(r_warnings, & &1.file, &{&1.position, &1.message})
-    compile_group = Enum.group_by(c_warnings, & &1.file, &{&1.position, &1.message})
+    runtime_group = Enum.group_by(r_warnings, & &1.file, &{&1.position, &1.message, &1[:span]})
+    compile_group = Enum.group_by(c_warnings, & &1.file, &{&1.position, &1.message, &1[:span]})
```

Comment:
Here is another place I think we should guarantee the span is always available, so we can use `&1.span`. Do you see any potential limitations for making the span key always present? :)

Reply:
You can grep for `position: ` and that should reveal most places where you are supposed to set the span. Also please change the type for `Code.diagnostic` :)

Reply:
Yeah, I guess I'll go and hunt for the places where the span key is not being added then :smile:

Reply:
Thank you for the tip, the grep on `position: ` was spot on :)

I think we should be good to go now, tests are passing here